src/transformStateWithClones.js

```diff
@@ -7,7 +7,60 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArr = [];
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+    const newState = stateArr.length
+      ? { ...stateArr[stateArr.length - 1] }
+      : { ...state };
+
+    switch (type) {
+      case 'addProperties':
+        stateArr.push(Object.assign(newState, extraData));
+        break;
+
+      case 'removeProperties':
+        if (!keysToRemove.length) {
```

Comment:
The check for `keysToRemove.length` is unnecessary. The `for...of` loop will not execute if `keysToRemove` is empty, so you can safely remove this check. [CODE STYLE]